From 50d500404645c62a96fd6fe5d0ef5111e7fe5f4e Mon Sep 17 00:00:00 2001
From: Ken Rockot <rockot@google.com>
Date: Wed, 20 Nov 2024 23:25:11 +0000
Subject: [PATCH] ipcz: Validate introduction transport

Gracefully handle (reject) an introduction containing an invalid
transport object.

This is not really an issue in production since these messages are only
accepted from brokers; but the fuzzer can trip over this.

Bug: 378917565
Change-Id: I8237bc1ae1da4318fa007bea3c6e6f21a6912b3e
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6037431
Reviewed-by: Alex Gough <ajgo@chromium.org>
Commit-Queue: Ken Rockot <rockot@google.com>
Cr-Commit-Position: refs/heads/main@{#1385948}
---

--- a/third_party/ipcz/src/ipcz/node_link.cc
+++ b/third_party/ipcz/src/ipcz/node_link.cc
@@ -474,8 +474,12 @@
   if (auto* v1 = accept.v1()) {
     remote_features = Features::Deserialize(accept, v1->remote_features);
   }
-  auto transport = MakeRefCounted<DriverTransport>(
-      accept.TakeDriverObject(accept.v0()->transport));
+  DriverObject transport_object =
+      accept.TakeDriverObject(accept.v0()->transport);
+  if (!transport_object.is_valid()) {
+    return false;
+  }
+  auto transport = MakeRefCounted<DriverTransport>(std::move(transport_object));
   node()->AcceptIntroduction(
       *this, accept.v0()->name, accept.v0()->link_side,
       accept.v0()->remote_node_type, accept.v0()->remote_protocol_version,
